isDDL and removeStatementHint panic with invalid query

It's possible to make things panic with an invalid query:

Failing examples are in https://github.com/googleapis/go-sql-spanner/pull/250

```
--- FAIL: FuzzIsDdl (0.21s)
    --- FAIL: FuzzIsDdl (0.00s)
        testing.go:1590: panic: runtime error: slice bounds out of range [:6] with length 0
            goroutine 121 [running]:
            runtime/debug.Stack()
                C:/Program Files/Go/src/runtime/debug/stack.go:24 +0x9b
            testing.tRunner.func1()
                C:/Program Files/Go/src/testing/testing.go:1590 +0x1c8
            panic({0x1ff99e0?, 0xc0001358c0?})
                C:/Program Files/Go/src/runtime/panic.go:770 +0x132
            github.com/googleapis/go-sql-spanner.isDDL({0x0?, 0x0?})
                s:/deps/go-sql-spanner/statement_parser.go:313 +0x1f8
            github.com/googleapis/go-sql-spanner.FuzzIsDdl.func1(0x0?, {0x0?, 0x0?})
                s:/deps/go-sql-spanner/statement_parser_test.go:870 +0x3b
```

And in `FuzzRemoveStatementHint`, with:

```
--- FAIL: FuzzRemoveStatementHint (6.19s)
    --- FAIL: FuzzRemoveStatementHint (0.00s)
        testing.go:1590: panic: runtime error: slice bounds out of range [:12] with length 10
            goroutine 3450 [running]:
            runtime/debug.Stack()
                C:/Program Files/Go/src/runtime/debug/stack.go:24 +0x9b
            testing.tRunner.func1()
                C:/Program Files/Go/src/testing/testing.go:1590 +0x1c8
            panic({0x1f599e0?, 0xc0003eaee8?})
                C:/Program Files/Go/src/runtime/panic.go:770 +0x132
            github.com/googleapis/go-sql-spanner.removeStatementHint({0xc00714c116, 0xa})
                s:/deps/go-sql-spanner/statement_parser.go:181 +0x35c
            github.com/googleapis/go-sql-spanner.FuzzRemoveStatementHint.func1(0x0?, {0xc00714c116?, 0x0?})
                s:/deps/go-sql-spanner/statement_parser_test.go:491 +0x3b
```

--- a/statement_parser.go
+++ b/statement_parser.go
@@ -177,9 +177,14 @@ func removeStatementHint(sql string) string {
 			break
 		}
 	}
-	if startQueryIndex > -1 {
+	// The startQueryIndex can theoretically be larger than the length of the SQL string,
+	// as the length of the uppercase SQL string can be different from the length of the
+	// lower/mixed case SQL string. This is however only the case for specific non-ASCII
+	// characters that are not allowed in a statement hint, so in that case we can safely
+	// assume the statement to be invalid.
+	if startQueryIndex > -1 && startQueryIndex < len(sql) {
 		endStatementHintIndex := strings.LastIndex(sql[:startQueryIndex], "}")
-		if startStatementHintIndex == -1 || startStatementHintIndex > endStatementHintIndex {
+		if startStatementHintIndex == -1 || startStatementHintIndex > endStatementHintIndex || endStatementHintIndex >= len(sql)-1 {
 			// Looks like an invalid statement hint. Just ignore at this point
 			// and let the caller handle the invalid query.
 			return sql
@@ -310,7 +315,7 @@ func isDDL(query string) (bool, error) {
 	// have already removed all leading spaces, and there are no keywords that
 	// start with the same substring as one of the DDL keywords.
 	for ddl := range ddlStatements {
-		if strings.EqualFold(query[:len(ddl)], ddl) {
+		if len(query) >= len(ddl) && strings.EqualFold(query[:len(ddl)], ddl) {
 			return true, nil
 		}
 	}

--- a/statement_parser_test.go
+++ b/statement_parser_test.go
@@ -372,8 +372,9 @@ SELECT 1`,
 
 func TestRemoveStatementHint(t *testing.T) {
 	tests := []struct {
-		input string
-		want  string
+		input              string
+		want               string
+		skipRemoveComments bool
 	}{
 		{
 			input: `@{JOIN_METHOD=HASH_JOIN} SELECT * FROM PersonsTable`,
@@ -451,11 +452,22 @@ SELECT SchoolID FROM Roster`,
 			input: `@JOIN_METHOD=HASH_JOIN SELECT * FROM PersonsTable`,
 			want:  `@JOIN_METHOD=HASH_JOIN SELECT * FROM PersonsTable`,
 		},
+		{
+			input:              "@{FORCE_INDEX=index_name}\xb0\xb0\xb0\x80SELECT",
+			want:               "@{FORCE_INDEX=index_name}\xb0\xb0\xb0\x80SELECT",
+			skipRemoveComments: true,
+		},
 	}
 	for _, tc := range tests {
-		sql, err := removeCommentsAndTrim(tc.input)
-		if err != nil {
-			t.Fatal(err)
+		var sql string
+		if tc.skipRemoveComments {
+			sql = tc.input
+		} else {
+			var err error
+			sql, err = removeCommentsAndTrim(tc.input)
+			if err != nil {
+				t.Fatal(err)
+			}
 		}
 		got := removeStatementHint(sql)
 		if got != tc.want {
@@ -793,6 +805,21 @@ func TestIsDdl(t *testing.T) {
 				A STRING(1024)
 			) PRIMARY KEY (A)`,
 			want: true,
+		},
+		{
+			name:  "short input",
+			input: "",
+			want:  false,
+		},
+		{
+			name:  "empty input",
+			input: "",
+			want:  false,
+		},
+		{
+			name:  "input with only spaces",
+			input: "    \t\n  ",
+			want:  false,
 		},
 	}
 

